src/backend/distributed/planner/multi_logical_planner.c

```diff
@@ -1580,43 +1580,33 @@ AllTargetExpressionsAreColumnReferences(List *targetEntryList)
 
 
 /*
- * RangeTableListContainsOnlyReferenceTables returns true if all range table
- * entries are reference tables.
- *
- * The function returns false for range table entries that are not relations.
- *
- * Note that the function doesn't recurse into subqueries, returns false when
- * a subquery is found.
+ * HasOnlyRecurringTuples returns if at least one of the relations in the given node
+ * is a distributed relation.
  */
 static bool
-RangeTableListContainsOnlyReferenceTables(List *rangeTableList)
+HasNonRecurringTuples(Node *node)
 {
+	List *rangeTableList = NULL;
 	ListCell *rangeTableCell = NULL;
+
+	/* extract range table entries for simple relations only */
+	ExtractRangeTableRelationWalker((Node *) node, &rangeTableList);
```

Comment:
Maybe:

```
bool
ContainsDistributedTableRTE(Node *node)
{
    return FindNodeCheck(node, IsDistributedTableRTE);
}

bool
IsDistributedTableRTE(Node *node)
{
    RangeTblEntry *rangeTableEntry = NULL;
    Oid relationId = InvalidOid;

    if (!IsA(node, RangeTblEntry))
    {
        return false;
    }

    rangeTableEntry = (RangeTblEntry *) node;
    if (rangeTableEntry->rtekind != RTE_RELATION)
    {
        return false;
    }

    relationId = rangeTableEntry->relid;
    if (!IsDistributedTable(relationId) ||
        PartitionMethod(relationId) == DISTRIBUTE_BY_NONE)
    {
        return false;
    }

    return true;
}
```

(copied from sql_monster)

Reply:
> Instead of constructing a big long list and then potentially only looking at the first item, maybe we could do:

makes sense

Reply:
Ah, that required some minor change to `FindNodeCheck()` given that if it doesn't start with a `Query *`, it doesn't recurse into the rte. But, at this point, we pass a `List *rtable`, which I've added a commit to fix

Reply:
Well, I think I prefer to leave this as is for now even it adds some overhead. It's much simpler than adding a new walker or modifying the existing ones.

Reply:
After this discussion, it appears to me that `ExtractRangeTableRelationWalker` has exponential complexity and produces duplicates.